### PR TITLE
routing: fix `index out of range` panic

### PR DESCRIPTION
**Pull request by positiveblue**
Make sure that we do not crash the node trying to access an invalid position for an array (-1) in `BuildRoute`.

Fixes https://github.com/lightningnetwork/lnd/issues/7798

### CODE REVIEW

File: routing/router.go [hunk 1]
```diff
@@ -2846,10 +2846,17 @@ func (r *ChannelRouter) BuildRoute(amt *lnwire.MilliSatoshi,
 	var pathEdges []*channeldb.CachedEdgePolicy
 	receiverAmt := runningAmt
 	for i, unifier := range unifiers {
+		var fromNode route.Vertex
```
Comment (guggero) [inline routing/router.go:2849]: Should we move this into the `if edge == nil {` branch to make it more explicit we're only using this in the error/not found case?
Also:
```
fromNode := source
if i > 0 {
    fromNode := hopts[i-1]
}
```
?

**Review by guggero — COMMENTED** [review]
Thanks for the fix!

**Review by guggero — APPROVED** [review]
LGTM :taco:

**Review by Crypt-iQ — APPROVED** [review]
lgtm pending ci 🦖

File: routing/router.go [hunk 2]
```diff
@@ -2848,8 +2848,13 @@ func (r *ChannelRouter) BuildRoute(amt *lnwire.MilliSatoshi,
 	for i, unifier := range unifiers {
 		edge := unifier.getEdge(receiverAmt, bandwidthHints)
 		if edge == nil {
+			fromNode := source
```
Comment (Roasbeef) [inline routing/router.go:2851]: Can we also add an itest to hit this case? 

Also is this the expected behavior? That the source becomes the default from node if only a single edge was specified?

**Review by Roasbeef — CHANGES_REQUESTED** [review]

**Review by Roasbeef — APPROVED** [review]
LGTM 🍾

**Pull request merged**